transport.go

```diff
@@ -82,3 +82,185 @@ var DefaultTransport = func(secure bool) (*http.Transport, error) {
 	}
 	return tr, nil
 }
+
+// transportTimeoutWrapper wraps a transport to time out on request and response body transfers.
+func transportTimeoutWrapper(tripper http.RoundTripper, timeout time.Duration) http.RoundTripper {
+	if timeout <= 0 {
+		return tripper
+	}
+	return &transportTimeout{parent: tripper, timeout: timeout}
+}
+
+type transportTimeout struct {
+	parent  http.RoundTripper
+	timeout time.Duration
+}
+
+func (t *transportTimeout) RoundTrip(request *http.Request) (*http.Response, error) {
+	request, ctx, cancel := newReqAliveChecker(request, t.timeout)
+	resp, err := t.parent.RoundTrip(request)
+	if err != nil {
+		cancel()
+		return resp, err
+	}
+	return newRespAliveChecker(ctx, cancel, resp, t.timeout), nil
+}
+
+// newReqAliveChecker will perform timeout checks between read calls.
+// This will measure time between body reads.
+// When the timeout is exceeded the request will be canceled.
+// A new request and context is returned.
+//
+func newReqAliveChecker(req *http.Request, timeout time.Duration) (request *http.Request, ctx context.Context, cancel context.CancelFunc) {
+	ctx, cancel = context.WithCancel(req.Context())
+	req = req.WithContext(ctx)
+
+	if req.Body == nil || timeout <= 0 {
+		return req, ctx, cancel
+	}
+	req.Body = &reqAliveChecker{timeout: timeout, rc: req.Body, ctx: ctx, cancel: cancel}
+	return req, ctx, cancel
+}
+
+type reqAliveChecker struct {
+	timedOut uint32
+	timeout  time.Duration
+	rc       io.ReadCloser
+
+	reset  chan struct{}
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (a *reqAliveChecker) Read(p []byte) (n int, err error) {
+	if atomic.LoadUint32(&a.timedOut) == 1 {
+		return 0, context.DeadlineExceeded
+	}
+	n, err = a.rc.Read(p)
+	// Start on first request
+	if err == nil {
+		if a.reset == nil {
+			update := make(chan struct{}, 1)
+			a.reset = update
+			go func() {
+				t := time.NewTimer(a.timeout)
```

Comment:
```suggestion
				t := time.NewTimer(a.timeout)
				defer t.Stop()
```

Missing Stop() upon return?